// @microsoft/azureportal-test/lib/Parts/ResourcePart/ResourcePart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../../Locators/By");
const Part_1 = require("../Part");
const ResourcePartProperty_1 = require("./ResourcePartProperty");
/**
 * Represents a ResourcePart.
 */
class ResourcePart extends Part_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxs-resource-part";
        this.resourcePartTitleClassName = "fxs-resource-part-title";
        this.resourcePartColumnClassName = "fxs-resource-part-column";
        /**
         * Gets columns.
         */
        this.columns = this.all(By_1.default.className(this.resourcePartColumnClassName));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.css(this.partCss), By_1.default.className(this.rootClassName));
    }
    /**
     * Gets properties contained in this part.
     *
     * @returns An ElementArrayFinder of ResourcePartProperty instances.
     */
    get properties() {
        return this.all(ResourcePartProperty_1.default);
    }
    /**
     * Gets the part title.
     */
    async getPartTitle() {
        return await this.element(By_1.default.className(this.resourcePartTitleClassName)).getText();
    }
    /**
     * Gets the part property that corresponds to the specified options.
     *
     * @param options The options used to find the part property.
     */
    property(options) {
        const locator = new ResourcePartProperty_1.default().buildLocator(options);
        return this.element(locator).asType(ResourcePartProperty_1.default);
    }
}
exports.default = ResourcePart;
