// @microsoft/azureportal-test/src/Parts/ButtonPart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const Part_1 = require("../Parts/Part");
/**
 * Represents a ButtonPart.
 */
class ButtonPart extends Part_1.default {
    constructor() {
        super(...arguments);
        this.buttonPartTitleClassName = "fxs-part-buttonpart-title";
        this.rootClassName = "fxs-part-buttonpart";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.css(this.partCss), By_1.default.className(this.rootClassName));
    }
    /**
     * @deprecated partTitle instance method is deprecated. Instead use getPartTitle instance method
     */
    async partTitle() {
        return await this.element(By_1.default.className(this.buttonPartTitleClassName)).getText();
    }
    /**
     * Gets the part title.
     */
    async getPartTitle() {
        return await this.element(By_1.default.className(this.buttonPartTitleClassName)).getText();
    }
}
exports.default = ButtonPart;
