// @microsoft/azureportal-test/src/Parts/CollectionPart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const Part_1 = require("./Part");
const Constants = require("../Constants");
const Grid_1 = require("../Controls/Grid");
/**
 * Represents a Collection part
 */
class CollectionPart extends Part_1.default {
    /**
     * @returns {Grid} The grid that shows the collection elements.
     */
    get grid() {
        return this.element(Grid_1.default);
    }
    /**
     * @returns {Promise} A promise that resolves with the rollup count of this CollectionPart
     */
    async getRollupCount() {
        const count = await this.element(By_1.default.className(Constants.CssClassNames.Parts.collectionPartRollupCount)).getText();
        return parseInt(count.trim());
    }
    /**
     * @returns {Promise} A promise that resolves with the rollup label of this CollectionPart
     */
    async getRollupLabel() {
        return await this.element(By_1.default.className(Constants.CssClassNames.Parts.collectionPartRollupLabel)).getText();
    }
    /**
     * @returns {Promise} A promise that resolves with the title of this CollectionPart
     */
    async getTitle() {
        return await this.element(By_1.default.className(Constants.CssClassNames.Parts.collectionPartTitle)).getText();
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns {Locator} A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.css(this.partCss), By_1.default.className(Constants.CssClassNames.Parts.collectionPart));
    }
}
exports.default = CollectionPart;
