// @microsoft/azureportal-test/lib/TopBar/ActionsMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionsMenu = void 0;
const DropMenu_1 = require("../Controls/DropMenu");
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
class ActionsMenu extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxs-topbar-actionsmenu";
        this.dropMenu = this.element(DropMenu_1.DropMenu);
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.rootClassName);
    }
    /**
     * Gets a value indicating whether the actions menu is opened.
     */
    async isOpened() {
        return await this.dropMenu.isOpened();
    }
    /**
     * Opens the actions menu if it isn't already opened.
     */
    async open() {
        return await this.dropMenu.open();
    }
    /**
     * Gets the the actions menu items by text.
     *
     * @param text the text of the actions menu item.
     */
    async getActionsMenuItem(text) {
        await this.open();
        return this.element(By_1.default.text(text));
    }
}
exports.ActionsMenu = ActionsMenu;
