// @microsoft/azureportal-test/lib/Notifications/ToastNotification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
const ElementArrayFinder_1 = require("../ElementArrayFinder");
/**
 * Represents the Toast Notification.
 */
class ToastNotification extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxs-toast-item";
        this.toastTitleClassName = "fxs-toast-title";
        this.toastCloseClassName = "fxs-toast-close";
        this.toastButtonClassName = "fxs-button";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.rootClassName);
    }
    /**
     * Gets the title associated to this element.
     *
     * @returns A instance.
     */
    get getToastTitle() {
        return this.element(By_1.default.className(this.toastTitleClassName));
    }
    /**
     * Get buttons in the toast.
     */
    async getToastButtons() {
        return this.all(By_1.default.className(this.toastButtonClassName));
    }
    /**
     * Click close button on toast.
     */
    async clickClose(dialogCb) {
        await ElementArrayFinder_1.default.element(By_1.default.className(this.toastCloseClassName)).click(undefined, dialogCb);
    }
}
exports.default = ToastNotification;
