// @microsoft/azureportal-test/lib/Utils/DefaultTimeouts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class DefaultTimeoutsImpl {
    constructor() {
        /**
         * The timeout for assert functions. Checks for user value otherwise defaults to 8 seconds.
         */
        this.assertTimeout = 8000;
        /**
         * The delay post blade wait methods. This is needed sometime when DOM is updated but rendering is slow.
         * Checks for user value otherwise defaults to 0.5 second.
         */
        this.bladePostWaitDelay = 500;
        /**
         * The default timeout used for delay calls.
         * Checks for user value otherwise defaults to 1 second.
         */
        this.delayTimeout = 1000;
        /**
         * The default extra long timeout used for actions that are expected to take more time than `longTimeout`
         * in CI (e.g. slower CloudTest agents).
         * Checks for user value otherwise defaults to 120 seconds.
         */
        this.extraLongTimeout = 120000;
        /**
         * Internal timeout used in ElementArrayFinder.waitForElementLocatedTimeout().
         * This timeout should never be longer than timeouts provided to portal.Wait (defaults to `timeout`) or portal.Wait may never retry.
         * Checks for user value otherwise defaults to 3 seconds.
         */
        this.internalWaitForElementLocatedTimeout = 3000;
        /**
         * The defaut long timeout used for actions that are expected to take longer than normal (e.g. loading the portal).
         * Checks for user value otherwise defaults to 60 seconds.
         */
        this.longTimeout = 60000;
        /**
         * The timeout used to check for a "page" event when getting a list of pages from browser (Playwright only).
         * Checks for user value otherwise defaults to 3 seconds.
         */
        this.pageEventTimeout = 3000;
        /**
         * The default timeout used across msportalfx-test.
         * Checks for user value otherwise defaults to 10 seconds.
         */
        this.timeout = 10000;
        /**
         * The timeout used to check whether url has settled. Checks for user value otherwise defaults to 30 seconds.
         */
        this.urlSettledTimeout = 30000;
    }
}
const DefaultTimeouts = new DefaultTimeoutsImpl();
Object.freeze(DefaultTimeouts);
exports.default = DefaultTimeouts;
