// @microsoft/azureportal-test/lib/Utils/Timeouts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pageEventTimeout = exports.urlSettledTimeout = exports.assertTimeout = exports.internalWaitForElementLocatedTimeout = exports.bladePostWaitDelay = exports.defaultExtraLongTimeout = exports.defaultLongTimeout = exports.defaultTimeout = void 0;
const DefaultTimeouts_1 = require("./DefaultTimeouts");
/**
 * @deprecated Use `DefaultTimeouts.timeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.defaultTimeout = DefaultTimeouts_1.default.timeout;
/**
 * @deprecated Use `DefaultTimeouts.longTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.defaultLongTimeout = DefaultTimeouts_1.default.longTimeout;
/**
 * @deprecated Use `DefaultTimeouts.extraLongTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.defaultExtraLongTimeout = DefaultTimeouts_1.default.extraLongTimeout;
/**
 * @deprecated Use `DefaultTimeouts.bladePostWaitDelay` instead.
 */
// eslint-disable-next-line prefer-const
exports.bladePostWaitDelay = DefaultTimeouts_1.default.bladePostWaitDelay;
/**
 * @deprecated Use `DefaultTimeouts.internalWaitForElementLocatedTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.internalWaitForElementLocatedTimeout = DefaultTimeouts_1.default.internalWaitForElementLocatedTimeout;
/**
 * @deprecated Use `DefaultTimeouts.assertTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.assertTimeout = DefaultTimeouts_1.default.assertTimeout;
/**
 * @deprecated Use `DefaultTimeouts.urlSettledTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.urlSettledTimeout = DefaultTimeouts_1.default.urlSettledTimeout;
/**
 * @deprecated Use `DefaultTimeouts.pageEventTimeout` instead.
 */
// eslint-disable-next-line prefer-const
exports.pageEventTimeout = DefaultTimeouts_1.default.pageEventTimeout;
