// @microsoft/azureportal-test/src/Controls/Dialogs/ModalDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../../Locators/By");
const DialogBase_1 = require("./DialogBase");
const MessageBox_1 = require("./MessageBox");
/**
 *  Represents a Modal dialog.
 */
class ModalDialog extends DialogBase_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxs-portal-shield-modaldialog";
        this.titleViewModelProperty = "title";
    }
    /**
     *  By locator used to identify elements of this class.
     *
     * @returns {Locator} A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.rootClassName);
    }
    /**
     * Gets the message box title.
     *
     * @returns Promise which resolves to modal dialog title.
     */
    async getTitle() {
        const selector = `[data-bind*=${this.titleViewModelProperty}]`;
        return await this.element(By_1.default.css(selector)).getText();
    }
    get messageBox() {
        return this.element(MessageBox_1.default);
    }
    /**
     * Perform the accept action on modal dialog.
     *
     * @param {string} buttonText The label of accept button.
     * @returns Promise which resolves to accept button is clicked.
     */
    async accept(buttonText = "OK") {
        await this.messageBox.clickButton(buttonText);
    }
}
exports.default = ModalDialog;
