// @microsoft/azureportal-test/src/Index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Key = exports.Dashboard = exports.Search = exports.until = exports.portal = exports.BrowserAutomation = exports.CoBrandPage = exports.BladeActionBar = exports.Lens = exports.ErrorPage = exports.CtrlAltDOperations = exports.TopBar = exports.StartBoard = exports.ElementArrayFinder = exports.SplashScreen = exports.PortalElement = exports.Portal = exports.LogLevel = exports.BrowserAutomationType = exports.WebDriverElement = exports.PortalWebDriver = exports.PlaywrightElement = exports.PortalPlaywright = exports.PortalConfigurationManager = exports.Views = exports.Logger = exports.Settings = exports.PortalFxResources = exports.Notifications = exports.Tests = exports.Utils = exports.Commands = exports.Parts = exports.Controls = exports.Constants = exports.ConfirmationAlertAction = exports.ContextPanes = exports.Browse = exports.Blades = exports.ActionBars = exports.Locators = void 0;
exports.Locators = require("./Locators");
exports.ActionBars = require("./ActionBars");
exports.Blades = require("./Blades");
exports.Browse = require("./Browse");
exports.ContextPanes = require("./ContextPanes");
exports.ConfirmationAlertAction = require("./ConfirmationAlertAction");
exports.Constants = require("./Constants");
exports.Controls = require("./Controls");
exports.Parts = require("./Parts");
exports.Commands = require("./Commands");
exports.Utils = require("./Utils");
exports.Tests = require("./Tests");
exports.Notifications = require("./Notifications");
exports.PortalFxResources = require("./PortalFxResources");
exports.Settings = require("./Settings");
exports.Logger = require("./Logger");
exports.Views = require("./Views");
exports.PortalConfigurationManager = require("./PortalConfigurationManager");
var PortalPlaywright_1 = require("./PortalPlaywright");
Object.defineProperty(exports, "PortalPlaywright", { enumerable: true, get: function () { return PortalPlaywright_1.PortalPlaywright; } });
Object.defineProperty(exports, "PlaywrightElement", { enumerable: true, get: function () { return PortalPlaywright_1.PlaywrightElement; } });
var PortalWebDriver_1 = require("./PortalWebDriver");
Object.defineProperty(exports, "PortalWebDriver", { enumerable: true, get: function () { return PortalWebDriver_1.PortalWebDriver; } });
Object.defineProperty(exports, "WebDriverElement", { enumerable: true, get: function () { return PortalWebDriver_1.WebDriverElement; } });
var BrowserAutomationType_1 = require("./BrowserAutomationType");
Object.defineProperty(exports, "BrowserAutomationType", { enumerable: true, get: function () { return BrowserAutomationType_1.BrowserAutomationType; } });
var LogLevel_1 = require("./LogLevel");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return LogLevel_1.LogLevel; } });
const Portal_1 = require("./Portal"), PortalModule = Portal_1;
exports.Portal = Portal_1.default;
const PortalElement_1 = require("./PortalElement");
exports.PortalElement = PortalElement_1.default;
const SplashScreen_1 = require("./SplashScreen");
exports.SplashScreen = SplashScreen_1.default;
const ElementArrayFinder_1 = require("./ElementArrayFinder");
exports.ElementArrayFinder = ElementArrayFinder_1.default;
const StartBoard_1 = require("./StartBoard");
exports.StartBoard = StartBoard_1.default;
const TopBar_1 = require("./TopBar/TopBar");
exports.TopBar = TopBar_1.default;
const CtrlAltDOperations_1 = require("./CtrlAltDOperations");
exports.CtrlAltDOperations = CtrlAltDOperations_1.default;
const ErrorPage_1 = require("./ErrorPage");
exports.ErrorPage = ErrorPage_1.default;
const Lens_1 = require("./Lens");
exports.Lens = Lens_1.default;
const BladeActionBar_1 = require("./ActionBars/BladeActionBar");
exports.BladeActionBar = BladeActionBar_1.default;
const CoBrandPage_1 = require("./CoBrandPage");
exports.CoBrandPage = CoBrandPage_1.default;
const BrowserAutomation = require("./BrowserAutomation");
exports.BrowserAutomation = BrowserAutomation;
const PlaywrightKeys = require("./PlaywrightKeys");
Object.defineProperty(module.exports, "portal", {
    get: function () {
        return PortalModule.portal;
    },
    set: function (value) {
        PortalModule.portal = value;
    },
});
exports.until = require("./ExpectedConditions");
exports.Search = require("./Search");
exports.Dashboard = require("./Dashboard");
/**
 * Representations of pressable keys that aren't text.
 * When using Selenium WebDriver, these are stored in the Unicode PUA (Private Use Area) code points, 0xE000-0xF8FF (refer to http://www.google.com.au/search?&q=unicode+pua&btnG=Search).
 * When using Playwright, these are stored in the KeyboardEvent.key values defined in PlaywrightKeys.ts (refer to https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values).
 */
const KeyObject = PlaywrightKeys;
exports.Key = new Proxy(KeyObject, {
    set: function () {
        return true;
    },
    get: function (_target, key) {
        return BrowserAutomation.getBrowser().Key[key];
    },
});
