// @microsoft/azureportal-test/src/PortalFxResources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectAll = exports.notifications = exports.untrustedPatch = exports.untrustedExtension = exports.yes = exports.usersMenu = exports.users = exports.subscription = exports.signInWithPassword = exports.signInWithUserAndPasswordInstead = exports.signInPageTitle = exports.settings = exports.selectExistingResourceGroup = exports.searchPlaceholder = exports.retry = exports.resourceGroups = exports.resourceGroup = exports.refresh = exports.quickStartMenu = exports.quickStart = exports.properties = exports.removeFromDashboard = exports.pinToDashboard = exports.pinnedToDashboard = exports.overview = exports.ok = exports.no = exports.nameLabel = exports.movingResourcesFailed = exports.moveResourceConfirm = exports.movingResourcesComplete = exports.movingResources = exports.moveResources = exports.move = exports.metrics = exports.marketplace = exports.manageView = exports.localDevelopment = exports.ignore = exports.createNew = exports.createANewGroup = exports.columns = exports.chooseColumnsV2 = exports.chooseColumns = exports.cancel = exports.browseAll = exports.browse = exports.allow = exports.abort = exports.activityLog = void 0;
exports.settingsFlyout = exports.frameworkClientMsPortalFxResources = exports.tagsClientStrings = exports.shellCommands = exports.hubsExtension = exports.debugHubResources = exports.ShellCommands = exports.dashboard = exports.unsafeURLParamsDialogContinue = exports.unsafeURLParamsDialogTitle = exports.listCommandDeleteListItem = exports.bladeTitleFormat = exports.edit = exports.browseResourceGroupsBladeTitle = exports.allResourcesBladeTitle = void 0;
exports.activityLog = "Activity log";
exports.abort = "Abort";
exports.allow = "Allow";
exports.browse = "Browse";
exports.browseAll = "Browse All";
exports.cancel = "Cancel";
exports.chooseColumns = "Choose columns";
exports.chooseColumnsV2 = "Edit columns";
exports.columns = "Edit columns";
exports.createANewGroup = "Create a new group";
exports.createNew = "Create new";
exports.ignore = "Ignore";
exports.localDevelopment = "Local Development";
exports.manageView = "Manage view";
exports.marketplace = "Marketplace";
exports.metrics = "Metrics";
exports.move = "Move";
exports.moveResources = "Move resources";
exports.movingResources = "Moving resources";
exports.movingResourcesComplete = "Moving resources complete";
exports.moveResourceConfirm = "I understand that tools and scripts associated with moved resources will not work until I update them to use new resource IDs";
exports.movingResourcesFailed = "Moving resources failed";
exports.nameLabel = "Name";
exports.no = "No";
exports.ok = "OK";
exports.overview = "Overview";
exports.pinnedToDashboard = "Pinned to dashboard.";
exports.pinToDashboard = "Pin to dashboard";
exports.removeFromDashboard = "Remove from dashboard";
exports.properties = "Properties";
exports.quickStart = "Quick start";
exports.quickStartMenu = "Quickstart";
exports.refresh = "Refresh";
exports.resourceGroup = "Resource group";
exports.resourceGroups = "Resource groups";
exports.retry = "Retry";
exports.searchPlaceholder = "Search resources";
exports.selectExistingResourceGroup = "Select existing resource group";
exports.settings = "Settings";
exports.signInPageTitle = "Sign in to Azure Portal";
exports.signInWithUserAndPasswordInstead = "Sign in with a username and password instead";
exports.signInWithPassword = "Password";
exports.subscription = "Subscription";
exports.users = "Users";
exports.usersMenu = "Access control (IAM)";
exports.yes = "Yes";
exports.untrustedExtension = "Untrusted Extensions!";
exports.untrustedPatch = "Untrusted Patch!";
exports.notifications = "Notifications";
exports.selectAll = "Select all";
exports.allResourcesBladeTitle = "All resources";
exports.browseResourceGroupsBladeTitle = "Resource groups";
exports.edit = "Edit";
exports.bladeTitleFormat = "{0} | {1}";
exports.listCommandDeleteListItem = "Delete";
exports.unsafeURLParamsDialogTitle = "This portal link is customized.";
exports.unsafeURLParamsDialogContinue = "Continue with overrides";
exports.dashboard = {
    resetToDefaultState: "Reset to default state",
    restoreDefaultLayout: "Restore default layout",
    create: "Create",
    blank: "Blank dashboard",
    delete: "Delete",
    share: "Share",
    edit: "Edit",
    manageSharing: "Manage sharing",
    unshare: "unshare",
    pinPane: {
        subscriptionLbl: "Subscription",
        dashboardNameLbl: "Dashboard name",
        createNew: "Create new",
        sharedRadioBtnTxt: "Shared",
        createAndPin: "Create and pin",
        pin: "Pin",
    },
    sharePane: {
        subscriptionLbl: "Subscription Name",
    },
    partGallery: {
        searchTiles: "Filter tiles",
    },
    clone: "Clone",
    cloneTitleFormat: "Clone of {0}",
    notFoundTitle: "Dashboard not found",
    download: "Download",
    export: "Export",
    upload: "Upload",
};
exports.ShellCommands = {
    partSizeMini: "1 x 1",
    partSizeSmall: "2 x 1",
    partSizeNormal: "2 x 2",
    partSizeWide: "4 x 2",
    partSizeTall: "2 x 4",
    partSizeHeroWide: "6 x 4",
    partSizeHeroTall: "4 x 6",
    partSizeHero: "6 x 6",
    partSizeLarge: "4 x 4",
};
exports.debugHubResources = {
    addPart: {
        buttonText: "Add tile",
    },
};
exports.hubsExtension = {
    extensionName: {
        name: "HubsExtension",
    },
    resourceGroups: {
        applyButton: "Apply",
        cancelButton: "Cancel",
        createButton: "Create",
        createTitle: "Create a resource group",
        deleteButton: "Delete",
        deleteConfirmation: "Are you sure you want to delete",
        locationLabel: "Region",
        nameLabel: "Resource group",
        nextReviewCreateButton: "Next : Review + create >",
        nextTagsButton: "Next : Tags >",
        noneTags: "None",
        resetButton: "Reset",
        reviewCreateButton: "Review + Create",
        tags: "Tags",
        templateAutomationLink: "Download a template for automation",
    },
    allResourcesBlade: {
        allResourcesBladeTitle: "All resources",
        refreshButtonText: "Refresh",
    },
    resourceGroupsBlade: {
        resourceGroupsBladeTitle: "Resource groups",
        refreshButtonText: "Refresh",
    },
    resourcePickerBlade: {
        resourcePickerBladeTitle: "Select resources",
        refreshButtonText: "Refresh",
    },
};
exports.shellCommands = {
    unpinFromStartboard: "Remove from dashboard",
    bladeRestore: "Restore",
};
exports.tagsClientStrings = {
    resourceTagsCollectionPart: {
        partTitle: "Tags",
    },
    resourceTagListsBlade: {
        bladeTitle: "Tags",
    },
    part: {
        settingList: {
            tags: "Tags",
        },
    },
    common: {
        save: "Save",
        applyLabel: "Apply",
        deleteAll: "Delete all",
    },
    assetTypeNames: {
        tag: {
            plural: "Tags",
        },
    },
    compoundFormat: "{0} : {1}",
};
exports.frameworkClientMsPortalFxResources = {
    editableGrid: {
        deleteCommandText: "Delete",
    },
};
exports.settingsFlyout = {
    appearanceAndStartupViewsTitle: "Appearance + startup views",
    contactInformation: "Contact information",
    directoriesTitle: "Directories + subscriptions",
    filtersTitle: "Advanced filters",
    overviewTitle: "Overview",
    languageAndRegionTitle: "Language + region",
    moreAzureResources: "More Azure resources",
    myInformationTitle: "My information",
    resetStateLabel: "Restore default settings",
    restore: "Restore",
    signoutAndNotificationsTitle: "Signing out + notifications",
};
