// @microsoft/azureportal-test/src/Parts/InfoList/InfoListItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../../Locators/By");
const PortalElement_1 = require("../../PortalElement");
/**
 * Represents a InfoListItem.
 */
class InfoListItem extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.itemTextClassName = "fxs-part-infolist-text";
        this.itemTextTitleClassName = "fxs-part-infolist-text-title";
        this.itemTextLinksClassName = "fxs-part-infolist-links";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.itemTextClassName);
    }
    /**
     * Gets the part title.
     */
    async partTitle() {
        return await this.element(By_1.default.className(this.itemTextTitleClassName)).getText();
    }
    /**
     * Gets the links element of the info list item.
     */
    getLinks() {
        return this.element(By_1.default.className(this.itemTextLinksClassName));
    }
}
exports.default = InfoListItem;
