// @microsoft/azureportal-test/src/Blades/BladeDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
/**
 * Represents a Blade Dialog.
 */
class BladeDialog extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.titleTag = "h2";
        /**
         * The dialog title.
         */
        this.title = this.element(By_1.default.tagName(this.titleTag));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className("fxs-blade-dialog");
    }
    /**
     * Clicks the specified button in the dialog.
     *
     * @param {string} buttonText The text of the button to click.
     * @returns A promise that resolves when the button has been clicked.
     */
    async clickButton(buttonText) {
        return await this.element(By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}*[(contains(@class, 'azc-button') or contains(@class, 'fxs-button')) and text()='${buttonText}']`)).click();
    }
    /**
     * Gets a field contained in this blade dialog that matches the given options.
     *
     * @param fieldType The type of field.
     * @param options The options used to locate the field.
     * @returns An instance of the field.
     */
    field(fieldType, options) {
        const fieldInstance = new fieldType();
        const locator = fieldInstance.buildLocator(options);
        return this.element(locator).asType(fieldType);
    }
    /**
     * Builds a BladeDialog locator using the specified options.
     *
     * @param options The options used to build the locator.
     * @returns A Locator instance.
     */
    buildLocator(options) {
        return options.title
            ? By_1.default.content(this.locator, By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}${this.titleTag}[text()="${options.title}"]`))
            : this.locator;
    }
}
exports.default = BladeDialog;
