// @microsoft/azureportal-test/lib/Blades/ContentState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentState = void 0;
/**
 * Represents the state of the content in a Part or a Blade.
 */
var ContentState;
(function (ContentState) {
    /**
     * No content state.
     */
    ContentState[ContentState["None"] = 0] = "None";
    /**
     * Success content state.
     */
    ContentState[ContentState["Success"] = 1] = "Success";
    /**
     * Warning content state.
     */
    ContentState[ContentState["Warning"] = 2] = "Warning";
    /**
     * Error content state
     */
    ContentState[ContentState["Error"] = 3] = "Error";
    /**
     * Dirty content state
     */
    ContentState[ContentState["Dirty"] = 4] = "Dirty";
    /**
     * Info content state
     */
    ContentState[ContentState["Info"] = 5] = "Info";
    /**
     * Upsell content state
     */
    ContentState[ContentState["Upsell"] = 6] = "Upsell";
    /**
     * Complete content state
     */
    ContentState[ContentState["Complete"] = 7] = "Complete";
})(ContentState || (exports.ContentState = ContentState = {}));
