// @microsoft/azureportal-test/lib/Blades/MoveResourcesBladeGrid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const GridBase = require("../Controls/GridBase");
/**
 * Represents the related resources grid inside the MoveResourcesBlade.
 */
class MoveResourcesBladeGrid extends GridBase.BaseGrid {
    constructor(elementArrayFinder) {
        super(MoveResourcesBladeGrid.Row, MoveResourcesBladeGrid.Cell, MoveResourcesBladeGrid.ColumnHeader, elementArrayFinder);
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className("ext-relatedresourcegrid");
    }
}
(function (MoveResourcesBladeGrid) {
    /**
     * Represents a column header in the MoveResourcesBladeGrid.
     */
    class ColumnHeader extends GridBase.BaseColumnHeader {
    }
    MoveResourcesBladeGrid.ColumnHeader = ColumnHeader;
    /**
     * Represents a cell in the MoveResourcesBladeGrid.
     */
    class Cell extends GridBase.BaseCell {
        constructor() {
            super(...arguments);
            /**
             * Gets the item name.
             */
            this.itemName = this.element(By_1.default.dataBindContains("text: settings.item.name"));
            /**
             * Gets the item type.
             */
            this.itemType = this.element(By_1.default.dataBindContains("text: settings.item.type"));
        }
    }
    MoveResourcesBladeGrid.Cell = Cell;
    /**
     * Represents a row in the MoveResourcesBladeGrid.
     */
    class Row extends GridBase.BaseRow {
        constructor(elementArrayFinder) {
            super(MoveResourcesBladeGrid.Cell, elementArrayFinder);
            /**
             * Gets the checkbox cell.
             */
            this.checkBoxCell = this.cells.get(0);
            /**
             * Gets the resource cell.
             */
            this.resourceCell = this.cells.get(1);
        }
    }
    MoveResourcesBladeGrid.Row = Row;
})(MoveResourcesBladeGrid || (MoveResourcesBladeGrid = {}));
exports.default = MoveResourcesBladeGrid;
