From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Kerr <charles@charleskerr.com>
Date: Tue, 6 Aug 2024 18:55:46 +0000
Subject: build: fix pdf

We don't want pdf, so don't include it.

diff --git a/chrome/browser/content_settings/BUILD.gn b/chrome/browser/content_settings/BUILD.gn
index 4130f3bd222983d433a54cc399dcf8f8dbe14ea6..1c844e8a4b5a1245b27e79f0e37301db54da0f6e 100644
--- a/chrome/browser/content_settings/BUILD.gn
+++ b/chrome/browser/content_settings/BUILD.gn
@@ -82,7 +82,6 @@ source_set("impl") {
     "//components/content_settings/core/browser:cookie_settings",
     "//components/pref_registry",
     "//components/supervised_user/core/browser",
-    "//pdf:features",
     "//services/metrics/public/cpp:ukm_builders",
     "//ui/webui",
   ]
diff --git a/chrome/browser/glic/BUILD.gn b/chrome/browser/glic/BUILD.gn
index f41defe8907835eed1ceaa61adaacf4cd6190bcd..023f6ec875a536ad4b1563769ca8b5c681fdc609 100644
--- a/chrome/browser/glic/BUILD.gn
+++ b/chrome/browser/glic/BUILD.gn
@@ -154,8 +154,6 @@ source_set("impl") {
     "//components/language/core/common",
     "//components/metrics_services_manager:metrics_services_manager",
     "//components/optimization_guide/proto:optimization_guide_proto",
-    "//components/pdf/browser:browser",
-    "//components/pdf/common:constants",
     "//components/sessions",
     "//extensions/browser:browser",
     "//ui/webui",
diff --git a/chrome/browser/ui/lens/BUILD.gn b/chrome/browser/ui/lens/BUILD.gn
index 87c78eac91f6b829453479b3ada880ae9459f5ae..dc9b674d57f69e8b8fee304507e903cbcd433081 100644
--- a/chrome/browser/ui/lens/BUILD.gn
+++ b/chrome/browser/ui/lens/BUILD.gn
@@ -239,7 +239,6 @@ source_set("browser_tests") {
     "//content/test:test_support",
     "//mojo/public/cpp/bindings",
     "//net",
-    "//pdf:features",
     "//third_party/lens_server_proto:lens_overlay_proto",
     "//ui/base",
     "//ui/base:types",
@@ -250,9 +249,6 @@ source_set("browser_tests") {
     "//ui/views/controls/webview",
     "//url",
   ]
-  if (enable_pdf) {
-    deps += [ "//chrome/browser/pdf" ]
-  }
 }
 
 source_set("interactive_ui_tests") {
@@ -279,7 +275,6 @@ source_set("interactive_ui_tests") {
     "//components/feature_engagement/public",
     "//components/lens:enterprise_policy",
     "//components/lens:features",
-    "//components/pdf/browser:browser",
     "//components/prefs",
     "//components/search_engines",
     "//components/user_education/views",
diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 692e44b75ac164a6f88398a4177913c47f49959e..ba601bf1db7280ef35e562a070127e5e8bca7260 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -3754,7 +3754,6 @@ if (!is_android) {
       deps += [
         "../browser/pdf",
         "../browser/pdf:pdf_extension_test_utils",
-        "../browser/ui/pdf",
         "//components/pdf/browser",
         "//components/pdf/common:constants",
         "//components/pdf/common:util",
