From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Kerr <charles@charleskerr.com>
Date: Tue, 6 Aug 2024 18:55:46 +0000
Subject: build: fix pdf

We don't want pdf, so don't include it.

diff --git a/chrome/browser/content_settings/BUILD.gn b/chrome/browser/content_settings/BUILD.gn
index 1fe1ecf20ad9fad8b1a37267c2dfafee1b4fc211..75e028ac25e43d37da0583689d2d6a87295115c7 100644
--- a/chrome/browser/content_settings/BUILD.gn
+++ b/chrome/browser/content_settings/BUILD.gn
@@ -83,7 +83,6 @@ source_set("impl") {
     "//components/content_settings/core/browser:cookie_settings",
     "//components/pref_registry",
     "//components/supervised_user/core/browser",
-    "//pdf:features",
     "//services/metrics/public/cpp:ukm_builders",
     "//ui/webui",
   ]
diff --git a/chrome/browser/glic/BUILD.gn b/chrome/browser/glic/BUILD.gn
index 69423f37f55fc3d4fa1a0a12693532f16981a810..74ab192d6ea269fa5493a9283cb852fb3e1c0c0c 100644
--- a/chrome/browser/glic/BUILD.gn
+++ b/chrome/browser/glic/BUILD.gn
@@ -196,8 +196,6 @@ source_set("impl") {
     "//components/metrics_services_manager:metrics_services_manager",
     "//components/optimization_guide/core",
     "//components/optimization_guide/proto:optimization_guide_proto",
-    "//components/pdf/browser:browser",
-    "//components/pdf/common:constants",
     "//components/pref_registry",
     "//components/prefs",
     "//components/sessions",
diff --git a/chrome/browser/ui/lens/BUILD.gn b/chrome/browser/ui/lens/BUILD.gn
index ce696eca19d69ac5e321a33dea1cce4db2ffdc65..184361f1182b21535053fb410ca416fdf06475c9 100644
--- a/chrome/browser/ui/lens/BUILD.gn
+++ b/chrome/browser/ui/lens/BUILD.gn
@@ -262,7 +262,6 @@ source_set("browser_tests") {
     "//content/test:test_support",
     "//mojo/public/cpp/bindings",
     "//net",
-    "//pdf:features",
     "//third_party/lens_server_proto:lens_overlay_proto",
     "//ui/base",
     "//ui/base:types",
@@ -273,9 +272,6 @@ source_set("browser_tests") {
     "//ui/views/controls/webview",
     "//url",
   ]
-  if (enable_pdf) {
-    deps += [ "//chrome/browser/pdf" ]
-  }
 }
 
 source_set("interactive_ui_tests") {
diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 16f2e321030d61ee2cbab91c15fc93da3cf735b5..88cb6f0abc6df52c2459b0f36dda8f389669eec8 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -3836,7 +3836,6 @@ if (!is_android) {
       deps += [
         "../browser/pdf",
         "../browser/pdf:pdf_extension_test_utils",
-        "../browser/ui/pdf",
         "//components/pdf/browser",
         "//components/pdf/common:constants",
         "//components/pdf/common:util",
