From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Kerr <charles@charleskerr.com>
Date: Tue, 6 Aug 2024 18:55:46 +0000
Subject: build: fix pdf

We don't want pdf, so don't include it.

diff --git a/chrome/browser/content_settings/BUILD.gn b/chrome/browser/content_settings/BUILD.gn
index e24de988473b298d952f2ec54982771c7b4cea72..33d7485944aab1df09197c24d352cb7afac3fe13 100644
--- a/chrome/browser/content_settings/BUILD.gn
+++ b/chrome/browser/content_settings/BUILD.gn
@@ -85,7 +85,6 @@ source_set("impl") {
     "//components/content_settings/core/browser:cookie_settings",
     "//components/pref_registry",
     "//components/supervised_user/core/browser",
-    "//pdf:features",
     "//services/metrics/public/cpp:ukm_builders",
     "//ui/webui",
   ]
diff --git a/chrome/browser/ui/lens/BUILD.gn b/chrome/browser/ui/lens/BUILD.gn
index 04eaccab86643bbad90efea0ed21e3a96e035177..345b9d7883fc90414ab2e7f3753850db59e0ce9a 100644
--- a/chrome/browser/ui/lens/BUILD.gn
+++ b/chrome/browser/ui/lens/BUILD.gn
@@ -292,7 +292,6 @@ source_set("browser_tests") {
     "//content/test:test_support",
     "//mojo/public/cpp/bindings",
     "//net",
-    "//pdf:features",
     "//third_party/lens_server_proto:lens_overlay_proto",
     "//ui/base",
     "//ui/base:types",
@@ -304,10 +303,6 @@ source_set("browser_tests") {
     "//ui/views/controls/webview",
     "//url",
   ]
-  if (enable_pdf) {
-    deps += [ "//chrome/browser/pdf" ]
-  }
-
   # TODO(crbug.com/40031409): Fix code that adds exit-time destructors and
   # enable the diagnostic by removing this line.
   configs += [ "//build/config/compiler:no_exit_time_destructors" ]
diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 4650080f45da4e321d4b0f0dabd98f2bf097c7c2..85d06b8c63a2798a070f4ce5635cd6f3da2a994e 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -4194,7 +4194,6 @@ if (!is_android) {
       deps += [
         "../browser/pdf",
         "../browser/pdf:pdf_extension_test_utils",
-        "../browser/ui/pdf",
         "//components/pdf/browser",
         "//components/pdf/common:constants",
         "//components/pdf/common:util",
