From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Kerr <charles@charleskerr.com>
Date: Tue, 6 Aug 2024 18:55:46 +0000
Subject: build: fix pdf

We don't want pdf, so don't include it.

diff --git a/chrome/browser/content_settings/BUILD.gn b/chrome/browser/content_settings/BUILD.gn
index a50a2a88b0420f9f740e56ac073b4967e50eedaa..49f251e6eeb3c5b4fde1eee37153e6b3866725f9 100644
--- a/chrome/browser/content_settings/BUILD.gn
+++ b/chrome/browser/content_settings/BUILD.gn
@@ -83,7 +83,6 @@ source_set("impl") {
     "//components/content_settings/core/browser:cookie_settings",
     "//components/pref_registry",
     "//components/supervised_user/core/browser",
-    "//pdf:features",
     "//services/metrics/public/cpp:ukm_builders",
     "//ui/webui",
   ]
diff --git a/chrome/browser/ui/lens/BUILD.gn b/chrome/browser/ui/lens/BUILD.gn
index 3b53b7954cbf22e4b54a76f62938ec61d915cf77..bd86302b534bd4aeba9953ad7c9194c616ac2577 100644
--- a/chrome/browser/ui/lens/BUILD.gn
+++ b/chrome/browser/ui/lens/BUILD.gn
@@ -279,7 +279,6 @@ source_set("browser_tests") {
     "//content/test:test_support",
     "//mojo/public/cpp/bindings",
     "//net",
-    "//pdf:features",
     "//third_party/lens_server_proto:lens_overlay_proto",
     "//ui/base",
     "//ui/base:types",
@@ -291,10 +290,6 @@ source_set("browser_tests") {
     "//ui/views/controls/webview",
     "//url",
   ]
-  if (enable_pdf) {
-    deps += [ "//chrome/browser/pdf" ]
-  }
-
   # TODO(crbug.com/40031409): Fix code that adds exit-time destructors and
   # enable the diagnostic by removing this line.
   configs += [ "//build/config/compiler:no_exit_time_destructors" ]
diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 244abec3d9a1758d78590f0f45b8af40615638e7..da581cb147dd379a15a83a0ecef4d713b80ab046 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -4084,7 +4084,6 @@ if (!is_android) {
       deps += [
         "../browser/pdf",
         "../browser/pdf:pdf_extension_test_utils",
-        "../browser/ui/pdf",
         "//components/pdf/browser",
         "//components/pdf/common:constants",
         "//components/pdf/common:util",
