From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Sun, 22 Nov 2020 14:07:01 -0800
Subject: Make autocorrect=off and spellcheck=false disable Touch Bar typing
 suggestions.

https://chromium-review.googlesource.com/c/chromium/src/+/1917603

diff --git a/content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm b/content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
index c89f25bf49d773f5f77b2434bd6fabcfd8cb4e33..066a13181fdbeeb5fc9bb754e7a8b18b69505a30 100644
--- a/content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
+++ b/content/app_shim_remote_cocoa/render_widget_host_view_cocoa.mm
@@ -2629,7 +2629,9 @@ - (void)invalidateTouchBar {
 }
 
 - (NSTouchBar*)makeTouchBar {
-  if (_textInputType != ui::TEXT_INPUT_TYPE_NONE) {
+  if (_textInputType != ui::TEXT_INPUT_TYPE_NONE &&
+      !(_textInputFlags & blink::kWebTextInputFlagAutocorrectOff) &&
+      !(_textInputFlags & blink::kWebTextInputFlagSpellcheckOff)) {
     _candidateListTouchBarItem = [[NSCandidateListTouchBarItem alloc]
         initWithIdentifier:NSTouchBarItemIdentifierCandidateList];
 
