// @microsoft/azureportal-test/src/Parts/ResourcePart/ResourcePartProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../../Locators/By");
const PortalElement_1 = require("../../PortalElement");
/**
 * Represents a property of a ResourcePart.
 */
class ResourcePartProperty extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.propertyClassName = "fxs-resource-part-property";
        this.nameClassName = "fxs-resource-part-property-name";
        this.valueClassName = "fxs-resource-part-property-value";
        /**
         * The name of the property.
         */
        this.name = this.element(By_1.default.className(this.nameClassName));
        /**
         * The value of the property.
         */
        this.value = this.element(By_1.default.className(this.valueClassName));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.propertyClassName);
    }
    /**
     * Builds a PartProperty locator using the specified options.
     *
     * @param options The options used to build the locator.
     * @returns A Locator instance.
     */
    buildLocator(options) {
        return options.name
            ? By_1.default.content(this.locator, By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}*[contains(@class, '${this.nameClassName}') and text()='${options.name}']`))
            : this.locator;
    }
}
exports.default = ResourcePartProperty;
