// @microsoft/azureportal-test/src/Controls/List/ListViewGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PortalElement_1 = require("../../PortalElement");
const By_1 = require("../../Locators/By");
const ListViewItem_1 = require("./ListViewItem");
/**
 * Represents a group in ListView.
 */
class ListViewGroup extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.groupClassName = "azc-listView-group";
        this.groupHeaderClassName = "azc-listView-groupheader";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.groupClassName);
    }
    /**
     * Gets all of the rows of the group.
     */
    get listViewItems() {
        return this.all(ListViewItem_1.default);
    }
    /**
     * Gets the header of the group.
     */
    get Header() {
        return this.element(By_1.default.className(this.groupHeaderClassName));
    }
    /**
     * Gets the specified item.
     *
     * @param identifier Specified identifier
     */
    getListViewItem(identifier) {
        return this.listViewItems
            .filter((listViewItem) => listViewItem.getText().then((text) => {
            text.toLocaleUpperCase().localeCompare(identifier) === 0;
        }))
            .first();
    }
}
exports.default = ListViewGroup;
