// @microsoft/azureportal-test/lib/PlaywrightKeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.F4 = exports.F3 = exports.F2 = exports.F1 = exports.DIVIDE = exports.DECIMAL = exports.SUBTRACT = exports.SEPARATOR = exports.ADD = exports.MULTIPLY = exports.NUMPAD9 = exports.NUMPAD8 = exports.NUMPAD7 = exports.NUMPAD6 = exports.NUMPAD5 = exports.NUMPAD4 = exports.NUMPAD3 = exports.NUMPAD2 = exports.NUMPAD1 = exports.NUMPAD0 = exports.EQUALS = exports.SEMICOLON = exports.DELETE = exports.INSERT = exports.DOWN = exports.ARROW_DOWN = exports.RIGHT = exports.ARROW_RIGHT = exports.UP = exports.ARROW_UP = exports.LEFT = exports.ARROW_LEFT = exports.HOME = exports.END = exports.PAGE_DOWN = exports.PAGE_UP = exports.SPACE = exports.ESCAPE = exports.PAUSE = exports.ALT = exports.CONTROL = exports.SHIFT = exports.ENTER = exports.RETURN = exports.CLEAR = exports.TAB = exports.BACK_SPACE = exports.HELP = exports.CANCEL = exports.NULL = void 0;
exports.META = exports.COMMAND = exports.F12 = exports.F11 = exports.F10 = exports.F9 = exports.F8 = exports.F7 = exports.F6 = exports.F5 = void 0;
// List of keyboard events identical to the one in Selenium WebDriver
exports.NULL = "";
exports.CANCEL = "Cancel";
exports.HELP = "Help";
exports.BACK_SPACE = "Backspace";
exports.TAB = "Tab";
exports.CLEAR = "Clear";
exports.RETURN = "Enter";
exports.ENTER = "Enter";
exports.SHIFT = "Shift";
exports.CONTROL = "Control";
exports.ALT = "Alt";
exports.PAUSE = "Pause";
exports.ESCAPE = "Escape";
exports.SPACE = " ";
exports.PAGE_UP = "PageUp";
exports.PAGE_DOWN = "PageDown";
exports.END = "End";
exports.HOME = "Home";
exports.ARROW_LEFT = "ArrowLeft";
exports.LEFT = "ArrowLeft";
exports.ARROW_UP = "ArrowUp";
exports.UP = "ArrowUp";
exports.ARROW_RIGHT = "ArrowRight";
exports.RIGHT = "ArrowRight";
exports.ARROW_DOWN = "ArrowDown";
exports.DOWN = "ArrowDown";
exports.INSERT = "Insert";
exports.DELETE = "Delete";
exports.SEMICOLON = ";";
exports.EQUALS = "=";
exports.NUMPAD0 = "0"; // number pad keys
exports.NUMPAD1 = "1";
exports.NUMPAD2 = "2";
exports.NUMPAD3 = "3";
exports.NUMPAD4 = "4";
exports.NUMPAD5 = "5";
exports.NUMPAD6 = "6";
exports.NUMPAD7 = "7";
exports.NUMPAD8 = "8";
exports.NUMPAD9 = "9";
exports.MULTIPLY = "Multiply";
exports.ADD = "Add";
exports.SEPARATOR = "Separator";
exports.SUBTRACT = "Subtract";
exports.DECIMAL = "Decimal";
exports.DIVIDE = "Divide";
exports.F1 = "F1"; // function keys
exports.F2 = "F2";
exports.F3 = "F3";
exports.F4 = "F4";
exports.F5 = "F5";
exports.F6 = "F6";
exports.F7 = "F7";
exports.F8 = "F8";
exports.F9 = "F9";
exports.F10 = "F10";
exports.F11 = "F11";
exports.F12 = "F12";
exports.COMMAND = "Meta"; // Apple command key
exports.META = "Meta"; // alias for Windows key
