// @microsoft/azureportal-test/lib/Controls/Button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PortalElement_1 = require("../PortalElement");
const By_1 = require("../Locators/By");
/**
 * Represents a button contained in an ActionBar.
 */
class Button extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        // The class of the button
        this.buttonClass = "fxs-button";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.css(`.${this.buttonClass}`);
    }
    buildLocator(options) {
        return By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}*[contains(@class, "${this.buttonClass}") and descendant::*[contains(@class, "fxs-button-text") and contains("${options.text}",text())]]`);
    }
    /**
     * Returns a value that indicates whether the button is enabled.
     *
     * @returns A promise that resolves with a value that indicates whether the button is enabled.
     */
    async isEnabled() {
        return !(await this.hasClass("fxs-button-disabled"));
    }
}
exports.default = Button;
