// @microsoft/azureportal-test/lib/Controls/Recommendations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
const Constants = require("../Constants");
/**
 * Represents a Recommendations.
 */
class Recommendations extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.emptyClass = "fxc-recommendations-empty";
        this.errorClass = "fxc-recommendations-general-error";
        this.bladeLink = "fxc-recommendations-blade-link";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(Constants.CssClassNames.Controls.recommendations);
    }
    /**
     * Gets blade link
     */
    getBladeLink() {
        return this.element(By_1.default.className(this.bladeLink));
    }
    /**
     * Gets a value indicating whether it is empty experience.
     */
    async isEmptyExperience() {
        return await this.element(By_1.default.className(this.emptyClass)).isDisplayed();
    }
    /**
     * Gets a value indicating whether it is error experience.
     */
    async isErrorExperience() {
        return await this.element(By_1.default.className(this.errorClass)).isDisplayed();
    }
}
exports.default = Recommendations;
