// @microsoft/azureportal-test/lib/Controls/SelectorField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const FormElement_1 = require("./FormElement");
/**
 * Represents a Selector field.
 */
class SelectorField extends FormElement_1.default {
    constructor() {
        super(...arguments);
        this.displayTextClassName = "fxc-selector-displayText";
        this.selectorChevronClassName = "fxc-selector-chevron";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className("azc-selectorField");
    }
    /**
     * Opens the associated picker by clicking on this field.
     *
     * @returns A promise that resolves when the the field has been clicked.
     */
    async openPicker() {
        return await this.click();
    }
    /**
     * Gets the display text.
     */
    async displayText() {
        return await this.element(By_1.default.className(this.displayTextClassName)).getText();
    }
    /**
     * Clicks on the chevron of the selector.
     */
    async click() {
        return await this.element(By_1.default.className(this.selectorChevronClassName)).click();
    }
}
exports.default = SelectorField;
