// @microsoft/azureportal-test/lib/Parts/AssetPart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const Part_1 = require("../Parts/Part");
/**
 * Represents a AssetPart.
 */
class AssetPart extends Part_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxs-part-assetpart";
        this.assetStatusClassName = "fxs-part-assetpart-status";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.css(this.partCss), By_1.default.className(this.rootClassName));
    }
    /**
     * Gets the asset name.
     */
    async getAssetName() {
        const assetElement = this.element(By_1.default.className("fxs-part-assetpart-assetName"));
        return await assetElement.getText();
    }
    /**
     * Gets the asset status.
     */
    async getStatus() {
        const assetElement = this.element(By_1.default.className(this.assetStatusClassName));
        return await assetElement.getText();
    }
}
exports.default = AssetPart;
