// @microsoft/azureportal-test/lib/Parts/PartProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
/**
 * Represents a property of a ResourceSummaryPart.
 */
class PartProperty extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.nameClass = "msportalfx-property-name";
        this.valueClass = "msportalfx-property-value";
        /**
         * The name of the property.
         */
        this.name = this.element(By_1.default.className(this.nameClass));
        /**
         * The value of the property.
         */
        this.value = this.element(By_1.default.className(this.valueClass));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className("msportalfx-property");
    }
    /**
     * Builds a PartProperty locator using the specified options.
     *
     * @param options The options used to build the locator.
     * @returns A Locator instance.
     */
    buildLocator(options) {
        return options.name
            ? By_1.default.content(this.locator, By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}*[contains(@class, '${this.nameClass}') and text()='${options.name}']`))
            : this.locator;
    }
}
exports.default = PartProperty;
