// @microsoft/azureportal-test/lib/Parts/ResourceSummaryPart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const Part_1 = require("./Part");
const PartProperty_1 = require("./PartProperty");
const resources = require("../PortalFxResources");
const HotSpot_1 = require("../Controls/HotSpot");
const QuickStartHotSpot_1 = require("../Controls/QuickStartHotSpot");
const AccessHotSpot_1 = require("../Controls/AccessHotSpot");
/**
 * Represents a ResourceSummary part.
 */
class ResourceSummaryPart extends Part_1.default {
    constructor() {
        super(...arguments);
        /**
         * The resource group property.
         */
        this.resourceGroupProperty = this.property({ name: resources.resourceGroup });
        /**
         * The Quick Start hotspot.
         */
        this.quickStartHotSpot = this.element(QuickStartHotSpot_1.default);
        /**
         * The Access hotspot.
         */
        this.accessHotSpot = this.element(AccessHotSpot_1.default);
        /**
         * The Settings hotspot.
         */
        this.settingsHotSpot = this.element(By_1.default.className("fxs-part-resourcesummary-settings")).asType(HotSpot_1.default);
        /**
         * The collection of part properties contained in this part.
         */
        this.properties = this.all(PartProperty_1.default);
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.css(this.partCss), By_1.default.className("fxs-part-resourcesummary"));
    }
    /**
     * Gets the part property that corresponds to the specified options.
     *
     * @param options The options used to find the part property.
     */
    property(options) {
        const locator = new PartProperty_1.default().buildLocator(options);
        return this.element(locator).asType(PartProperty_1.default);
    }
}
exports.default = ResourceSummaryPart;
