// @microsoft/azureportal-test/lib/Parts/TileSize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TileSize = void 0;
/**
 * Represents the possible sizes that Tile can have.
 */
var TileSize;
(function (TileSize) {
    /**
     * The mini size
     */
    TileSize[TileSize["Mini"] = 0] = "Mini";
    /**
     * The small size
     */
    TileSize[TileSize["Small"] = 1] = "Small";
    /**
     * The normal size
     */
    TileSize[TileSize["Normal"] = 2] = "Normal";
    /**
     * The wide size
     */
    TileSize[TileSize["Wide"] = 3] = "Wide";
    /**
     * The tall size
     */
    TileSize[TileSize["Tall"] = 4] = "Tall";
    /**
     * The hero wide size
     */
    TileSize[TileSize["HeroWide"] = 5] = "HeroWide";
    /**
     * The hero tall size
     */
    TileSize[TileSize["HeroTall"] = 6] = "HeroTall";
    /**
     * The hero wide fit height size
     */
    TileSize[TileSize["HeroWideFitHeight"] = 7] = "HeroWideFitHeight";
    /**
     *  The full width fit height size
     */
    TileSize[TileSize["FullWidthFitHeight"] = 8] = "FullWidthFitHeight";
    /**
     * The fit to container size
     */
    TileSize[TileSize["FitToContainer"] = 9] = "FitToContainer";
    /**
     * The large size
     */
    TileSize[TileSize["Large"] = 10] = "Large";
    /**
     * The hero size
     */
    TileSize[TileSize["Hero"] = 11] = "Hero";
    /**
     * The custom size
     */
    TileSize[TileSize["Custom"] = 99] = "Custom";
})(TileSize || (exports.TileSize = TileSize = {}));
