// @microsoft/azureportal-test/lib/Locators/ContentLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Locator_1 = require("./Locator");
/**
 * Represents a Locator composed of other locators. The findElements function returns the elements found by the
 * first locator provided in the constructor.
 */
class ContentLocator extends Locator_1.default {
    /**
     * Creates a new instance of this class.
     *
     * @param {Locator[]} innerLocators The collection of locators used to find elements.
     */
    constructor(innerLocators) {
        super();
        this.locators = new Array();
        innerLocators.forEach((locator) => {
            this.locators.push(locator);
        });
    }
    /**
     * Schedules a command to search for multiple elements on the page.
     *
     * @param {IBrowserAutomation | IBrowserElement} context The object used as a starting point for the search.
     * @returns An array of the webdriver.WebElement objects that were found.
     */
    findElements(context, waitUntilLocated, timeout) {
        return this.findElementsInternal(context, 0, null, waitUntilLocated, timeout).then((elements) => {
            return elements;
        });
    }
    /**
     * Gets the string representation of this locator.
     *
     * @returns The string representation of this locator.
     */
    toString() {
        let locatorString;
        this.locators.forEach((locator) => {
            if (!locatorString) {
                locatorString = locator.toString();
            }
            else {
                locatorString += ", " + locator.toString();
            }
        });
        return `By.Content(${locatorString})`;
    }
    async findElementsInternal(context, currentIndex, rootElements, waitUntilLocated, timeout) {
        if (!rootElements) {
            return this.locators[currentIndex].findElements(context).then((elements) => {
                if (currentIndex + 1 === this.locators.length) {
                    return Promise.resolve(elements);
                }
                else {
                    return this.findElementsInternal(context, currentIndex + 1, elements);
                }
            });
        }
        else {
            const innerPromises = new Array();
            const currentIndexLocator = this.locators[currentIndex];
            rootElements.forEach((element) => {
                const innerPromise = currentIndexLocator
                    .findElements(element, waitUntilLocated, timeout)
                    .then((elements) => {
                    if (elements && elements.length > 0) {
                        if (currentIndex + 1 === this.locators.length) {
                            return element;
                        }
                        else {
                            return this.findElementsInternal(context, currentIndex + 1, elements).then((foundElements) => {
                                return foundElements.length > 0 ? element : null;
                            });
                        }
                    }
                });
                innerPromises.push(innerPromise);
            });
            const elements = await Promise.all(innerPromises);
            const returnedElements = new Array();
            elements.forEach((element) => {
                if (element) {
                    returnedElements.push(element);
                }
            });
            return returnedElements;
        }
    }
}
exports.default = ContentLocator;
