// @microsoft/azureportal-test/lib/Controls/List/ListViewItem2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PortalElement_1 = require("../../PortalElement");
const By_1 = require("../../Locators/By");
/**
 * Represents a list view item.
 */
class ListViewItem2 extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.listItemClassName = "fxc-listView-item";
        this.selectedItem = "aria-selected";
        this.currentItem = "aria-current";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.listItemClassName);
    }
    /**
     * Gets a value indicating whether or not this element is selected.
     */
    async isSelected() {
        const current = await this.getAttribute(this.currentItem);
        if (current === "true") {
            return true;
        }
        else {
            const selected = await this.getAttribute(this.selectedItem);
            return selected === "true";
        }
    }
}
exports.default = ListViewItem2;
