// @microsoft/azureportal-test/src/Controls/CopyableLabel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const FormElement_1 = require("./FormElement");
const Constants = require("../Constants");
/**
 * Represents a CopyableLabel.
 */
class CopyableLabel extends FormElement_1.default {
    constructor() {
        super(...arguments);
        /**
         * The input element.
         */
        this.input = this.element(By_1.default.tagName("input"));
        /**
         * The copy button element.
         */
        this.copyButton = this.element(By_1.default.className(Constants.CssClassNames.Controls.copyableLabelCopyButton));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(Constants.CssClassNames.Controls.copyableLabel);
    }
    /**
     * Gets the current value of the CopyableLabel
     *
     * @returns A promise that will resolve with the current value of the CopyableLabel
     */
    async value() {
        return await this.input.getAttribute("value");
    }
    /**
     * Enters the specified text in the CopyableLabel.
     *
     * @param var_args The text to enter.
     * @returns A promise that resolves when the text has been entered.
     */
    async sendKeys(...var_args) {
        // eslint-disable-next-line prefer-spread
        return await this.input.sendKeys.apply(this.input, var_args);
    }
    /**
     * Schedules a command to clear the value of this element.
     *
     * @returns A promise that will be resolved when the element has been cleared.
     */
    async clear() {
        return await this.input.clear.apply(this.input);
    }
}
exports.default = CopyableLabel;
