// @microsoft/azureportal-test/src/Controls/CreatorAndSelectorField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const TextField_1 = require("./TextField");
const FormElement_1 = require("./FormElement");
const Logger_1 = require("../Logger");
/**
 * Represents a CreatorAndSelector field.
 */
class CreatorAndSelectorField extends FormElement_1.default {
    constructor() {
        super(...arguments);
        this.createNewTextField = this.element(TextField_1.default);
        this.createNewLink = this.element(By_1.default.css(".fxc-selector-link"));
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className("azc-creatorAndSelectorField");
    }
    /**
     * Opens the associated picker by clicking on this field.
     *
     * @returns A promise that resolves when the the field has been clicked.
     */
    async openPicker() {
        Logger_1.default.information("Clicking creator and selector...");
        return await this.click();
    }
    /**
     * Clicks the Create New link on this field.
     *
     * @returns A promise that resolves when the Create New link has been clicked.
     */
    async clickCreateNew() {
        Logger_1.default.information("Clicking 'Create new' in creator and selector...");
        return await this.createNewLink.click();
    }
    /**
     * Enters a new value in the creator TextField.
     *
     * @param var_args Text to send to the TextField.
     * @returns A promise that resolves with this CreatorAndSelectorField instance
     * when the text has been entered in the creator TextField.
     */
    async enterNewValue(...var_args) {
        const argsArray = [];
        var_args.forEach((arg) => {
            argsArray.push(arg);
        });
        // eslint-disable-next-line prefer-spread
        await this.createNewTextField.sendKeys.apply(this.createNewTextField, argsArray);
        return this;
    }
}
exports.default = CreatorAndSelectorField;
