// @microsoft/azureportal-test/src/Controls/MonitorChart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
const Constants = require("../Constants");
/**
 * Represents a Monitor Chart v2 part.
 */
class MonitorChart extends PortalElement_1.default {
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.content(By_1.default.className(Constants.CssClassNames.Controls.monitorChart));
    }
    /**
     * Gets whether the chart has data
     */
    async hasData() {
        return await this.isElementPresent(By_1.default.className(Constants.CssClassNames.Controls.monitorChartData));
    }
    /**
     * Gets the name of the monitor chart's first metric.
     */
    async getMetric() {
        return await this.element(By_1.default.className(Constants.CssClassNames.Controls.monitorChartMetric)).getText();
    }
    /**
     * Gets the value of the monitor chart's first metric.
     */
    async getValue() {
        return await this.element(By_1.default.className(Constants.CssClassNames.Controls.monitorChartValue)).getText();
    }
    /**
     * Gets the unit of the monitor chart's first metric.
     */
    async getUnit() {
        return await this.element(By_1.default.className(Constants.CssClassNames.Controls.monitorChartUnit)).getText();
    }
}
exports.default = MonitorChart;
