// @microsoft/azureportal-test/src/Controls/SimpleButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PortalElement_1 = require("../PortalElement");
const By_1 = require("../Locators/By");
/**
 * Represents a simple button.
 */
class SimpleButton extends PortalElement_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = "fxc-simplebutton";
        this.buttonClassName = "fxs-button";
        this.disabledClassName = "fxs-button-disabled";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.className(this.rootClassName);
    }
    buildLocator(options) {
        return By_1.default.xpath(`${By_1.default.currentNodeXPathPrefix()}*[contains(@class, "${this.buttonClassName}") and descendant::*[contains(@class, "fxs-button-text") and contains("${options.text}",text())]]`);
    }
    /**
     * Checks if this button is enabled
     *
     * @returns True, if enabled. False, otherwise
     */
    async isEnabled() {
        return !((await this.hasClass(this.disabledClassName)) ||
            (await this.element(By_1.default.className(this.disabledClassName)).isPresent()));
    }
}
exports.default = SimpleButton;
