// @microsoft/azureportal-test/src/ActionBars/FormActionBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const By_1 = require("../Locators/By");
const PortalElement_1 = require("../PortalElement");
const ActionBarButton_1 = require("./ActionBarButton");
/**
 * Represents an FormActionBar.
 */
class FormActionBar extends PortalElement_1.default {
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.css(".fxc-formActionBar");
    }
    /**
     * Gets the save button from Form Action Bar.
     *
     * @returns A Save button.
     */
    get saveButton() {
        return this.element(FormActionBarSaveButton).asType(ActionBarButton_1.default);
    }
    /**
     * Gets the discard button from Form Action Bar.
     *
     * @returns A Discard button.
     */
    get discardButton() {
        return this.element(FormActionBarDiscardButton).asType(ActionBarButton_1.default);
    }
}
exports.default = FormActionBar;
/**
 * Represents an Form Action Bar Save button.
 */
class FormActionBarSaveButton extends ActionBarButton_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = ".fxc-formActionBar-saveButton";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.css(this.rootClassName);
    }
}
/**
 * Represents an Form Action Bar Discard button.
 */
class FormActionBarDiscardButton extends ActionBarButton_1.default {
    constructor() {
        super(...arguments);
        this.rootClassName = ".fxc-formActionBar-discardButton";
    }
    /**
     * Gets the locator associated to this element.
     *
     * @returns A Locator instance.
     */
    get locator() {
        return By_1.default.css(this.rootClassName);
    }
}
